port/linux/fw_update_linux.c

```diff
@@ -67,4 +84,56 @@ golioth_status_t fw_update_change_boot_image(void) {
     return GOLIOTH_OK;
 }
 
-void fw_update_end(void) {}
+void fw_update_end(void) {
+    if (_filebuf) {
+        free(_filebuf);
+        _filebuf = NULL;
+    }
+    _initialized = false;
+}
+
+golioth_status_t fw_update_read_current_image_at_offset(
+        uint8_t* buf,
+        size_t bufsize,
+        size_t offset) {
+    static int filesize = 0;
+
+    if (!_initialized) {
+        char current_exe_path[256];
+
+        // Get the path to the current running executable
+        ssize_t nbytes = readlink("/proc/self/exe", current_exe_path, sizeof(current_exe_path));
+
+        if (nbytes > 0) {
+            GLTH_LOGI(TAG, "Current exe: %s", current_exe_path);
+        } else {
+            GLTH_LOGE(TAG, "Error getting path to current executable: %ld", nbytes);
+            return GOLIOTH_ERR_FAIL;
+        }
+
+        // read the entire file into a buffer
+        int fd;
+        if (((fd = open(current_exe_path, O_RDONLY, 0)) < 0)
+            || ((filesize = lseek(fd, 0, SEEK_END)) == -1)
+            || ((_filebuf = malloc(filesize + 1)) == NULL) || (lseek(fd, 0, SEEK_SET) != 0)
+            || (read(fd, _filebuf, filesize) != filesize) || (close(fd) == -1)) {
+            GLTH_LOGE(TAG, "error reading file");
+        }
```

Comment:
There are way too many operations  in a single `if` statement :)

Reply:
You're right. I copied this snippet from the `bsdiff` test code, just to get something working quickly.

I'll refactor it into a `read_file()` function.